### PR TITLE
convert penaltykickplay to playfsm

### LINKED ISSUE #2747: Convert PenaltyKickPlay into using PlayFSMs
### Description of the task

Some of our plays are written in Coroutines, other plays are written in boost::SML. We want to transition to using boost::SML so that we can write more readable gameplay and write gameplay that re-uses previously written code.

In this issue, we will convert PenaltyKickPlay into using a PlayFSM (no need to touch PenaltyKickEnemyPlay). You may try to do a 1-1 port of the existing C++ gameplay, or you can write it according to your vision of how penalty kicks should work and change the approach.

You may find it helpful reading up on how the league sends RefereeCommands to set up a PenaltyKick: [SSL Rules
](https://robocup-ssl.github.io/ssl-rules/sslrules.html#_penalty_kick)

If you want to see an example of how PlayFSMs are made, take a look at ShootOrPassPlayFSM or CreaseDefencePlayFSM.
Another resource is TacticFSM, you can find many of those in the `software/ai/hl/stp/tactic/ `folder

Currently, the PenaltyKickPlay tests are disabled (https://github.com/UBC-Thunderbots/Software/blob/master/src/software/ai/hl/stp/play/penalty_kick_play_test.cpp) because there’s a dribbling bug that causes the robot to occasionally lose the ball (fixing the dribbling is not part of this Issue, you can make the port and we can fix dribbling separately in another issue). If you want to re-enable these tests change `DISABLED_test_penalty_kick_take` -> `test_penalty_kick_take` and `DISABLED_test_penalty_kick_setup` into `test_penalty_kick_setup`

### Acceptance criteria

- [ ] Create a `penalty_kick_play_fsm.h` and `penalty_kick_play_fsm.cpp` that returns 5 move tactics + 1 PenaltyKickSetup tactic to set up the penalty kick
- [ ] Handle the transition into taking a penalty and return 5 move tactics + 1 PenaltyKick Tactic to approach the goalie and shoot
- [ ] Add a control params so that we can adjust how many move tactics we return
- [ ] Create a `penalty_kick_fsm_test.cpp` and test the transitions for the play you made
- [ ] Modify `penalty_kick_play.cpp` to use the PenaltyKickFSM

### PR DESCRIPTION
**Pull request by germchow**
### Description

Converts the penalty kick play to a FSM (originally used coroutines)

(Does not include fix for penalty kick tactic which is currently buggy: #2793)

### Testing Done

Behaves effectively the same as the previous penalty kick play implementation

### Resolved Issues

resolves #2747

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/software/ai/hl/stp/play/penalty_kick_play_fsm.cpp [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    return event.common.world.gameState().isStopped();
+}
+
+void PenaltyKickPlayFSM::setPositioningTactics()
+{
+    penalty_setup_tactics = std::vector<std::shared_ptr<PenaltySetupTactic>>(5);
+    std::generate(penalty_setup_tactics.begin(), penalty_setup_tactics.end(),
+                  [this]() { return std::make_shared<PenaltySetupTactic>(); });
+
+    away_from_kick_tactics = std::vector<std::shared_ptr<StopTactic>>(4);
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/penalty_kick_play_fsm.cpp]: Ideally we want this to be scalable to div A and also work if some of our robots have fouled out. So let's try getting this number from World and seeing how many friendly robots we have

File: src/software/ai/hl/stp/play/penalty_kick_play_fsm.h [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+    /**
+     * Action to make the penalty kick taking robot perform the PenaltyKickTactic
+     *
+     * @param event the PenaltyKickPlayFSM Update event
+     */
+    void performKick(const Update& event);
+
+    /**
+     * Guard to check if a pass has been found
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/penalty_kick_play_fsm.h]: what do you mean by "pass"?
do you mean "shot" here?

**Review by itsarune — COMMENTED** [review]
well done learning and making this FSM!

some comments:
- could you add a penalty_kick_play_fsm_test.cpp to test a basic transitions (as a sanity check)
- could you reorganize the new files under a penalty_kick folder src/software/ai/hl/stp/play/penalty_kick

File: src/software/ai/hl/stp/play/penalty_kick_play_fsm.cpp [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        Point(ball_position_x - 1.25, 0),
+        event.common.world.field().enemyGoalCenter().toVector().orientation(), 0);
+    penalty_setup_tactics.at(1)->updateControlParams(
+        Point(ball_position_x - 1.25, 4 * ROBOT_MAX_RADIUS_METERS),
+        event.common.world.field().enemyGoalCenter().toVector().orientation(), 0);
+    penalty_setup_tactics.at(2)->updateControlParams(
+        Point(ball_position_x - 1.25, -4 * ROBOT_MAX_RADIUS_METERS),
+        event.common.world.field().enemyGoalCenter().toVector().orientation(), 0);
+    penalty_setup_tactics.at(3)->updateControlParams(
+        Point(ball_position_x - 1.25, 8 * ROBOT_MAX_RADIUS_METERS),
```
Comment (williamckha) [inline src/software/ai/hl/stp/play/penalty_kick_play_fsm.cpp]: should 1.25 be a constant? Might also be clearer to have `double behind_penalty_mark_x = ball_position_x - 1.25` and plug that into the points

**Review by williamckha — COMMENTED** [review]
LGTM, just curious

**Review by williamckha — DISMISSED** [review]

File: src/software/ai/hl/stp/play/penalty_kick_play_fsm.h [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        DEFINE_SML_EVENT(Update)
+
+        DEFINE_SML_ACTION(setupPosition)
+        DEFINE_SML_ACTION(performKick)
+
+        DEFINE_SML_GUARD(setupPositionDone)
+        DEFINE_SML_GUARD(kickDone)
+
+        return make_transition_table(
+            // src_state + event [guard] / action = dest_state
```
Comment (JoshuaWu7) [inline src/software/ai/hl/stp/play/penalty_kick_play_fsm.h]: Is there a possibility to embed a fsm diagram as ASCII art? Or have a copy of the penaltykickplay as a fsm diagram?

File: src/software/ai/hl/stp/play/penalty_kick_play.cpp [hunk 5]
```diff
@@ -1,74 +1,30 @@
...
-        // If we are setting up for penalty kick, move our robots to position
-        if (world.gameState().isSetupState())
-        {
-            tactics_to_run[0].emplace_back(shooter_setup_move);
-        }
-        else if (world.gameState().isOurPenalty())
-        {
-            tactics_to_run[0].emplace_back(penalty_shot_tactic);
-        }
+    // This function doesn't get called so it does nothing
```
Comment (JoshuaWu7) [inline src/software/ai/hl/stp/play/penalty_kick_play.cpp]: This function can be removed if it does nothing?

File: src/software/ai/hl/stp/play/penalty_kick_play_fsm.cpp [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+                         event.common.world.ball().position())
+                            .orientation();
+
+    Point behind_ball = event.common.world.ball().position() +
+                        behind_ball_direction.normalize(DIST_TO_FRONT_OF_ROBOT_METERS +
+                                                        BALL_MAX_RADIUS_METERS + 0.1);
+    double ball_position_x = event.common.world.field().friendlyPenaltyMark().x();
+
+    unsigned int num_tactics = event.common.num_tactics;
+    if (num_tactics != away_from_kick_tactics.size())
```
Comment (JoshuaWu7) [inline src/software/ai/hl/stp/play/penalty_kick_play_fsm.cpp]: Maybe add a comment to explain what this if statement block is doing?

**Review by JoshuaWu7 — COMMENTED** [review]
Nice work on converting penaltykickplay to a fsm, just left some comments on the PR

**Review by williamckha — DISMISSED** [review]

File: src/software/ai/hl/stp/play/penalty_kick_play.cpp (hunk 5 above)
Comment (JoshuaWu7) [reply src/software/ai/hl/stp/play/penalty_kick_play.cpp]: Leave comment to remove later once all coroutines are removed from play abstract class?

File: src/software/ai/hl/stp/play/penalty_kick_play_fsm.cpp (hunk 6 above)
Comment (JoshuaWu7) [reply src/software/ai/hl/stp/play/penalty_kick_play_fsm.cpp]: bump

**Review by JoshuaWu7 — CHANGES_REQUESTED** [review]

File: src/software/ai/hl/stp/play/penalty_kick_play_fsm.cpp (hunk 6 above)
Comment (williamckha) [reply src/software/ai/hl/stp/play/penalty_kick_play_fsm.cpp]: I've seen a similar code block grouped into its own helper function (e.g. in CreaseDefensePlayFSM) which better describes what this code does

https://github.com/UBC-Thunderbots/Software/blob/c070809816e7711d3e1d65d4a1aef0a54d866bb7/src/software/ai/hl/stp/play/crease_defense/crease_defense_play_fsm.h#L37-L42

https://github.com/UBC-Thunderbots/Software/blob/c070809816e7711d3e1d65d4a1aef0a54d866bb7/src/software/ai/hl/stp/play/crease_defense/crease_defense_play_fsm.cpp#L41-L48

**Review by itsarune — COMMENTED** [review]
all good imo, pending Joshua's comment

**Review by williamckha — DISMISSED** [review]

**Review by williamckha — DISMISSED** [review]

File: src/software/ai/hl/stp/play/penalty_kick_play.cpp (hunk 5 above)
Comment (JoshuaWu7) [reply src/software/ai/hl/stp/play/penalty_kick_play.cpp]: Discussed offline already, we can't remove couroutine methods yet.

File: src/software/ai/hl/stp/play/penalty_kick_play_fsm.cpp (hunk 6 above)
Comment (JoshuaWu7) [reply src/software/ai/hl/stp/play/penalty_kick_play_fsm.cpp]: This helper function is non-blocking for me, resolving comment.

**Review by JoshuaWu7 — DISMISSED** [review]
LGTM

File: src/software/ai/hl/stp/play/penalty_kick/penalty_kick_play_fsm.cpp [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+                            .orientation();
+
+    Point behind_ball = event.common.world.ball().position() +
+                        behind_ball_direction.normalize(DIST_TO_FRONT_OF_ROBOT_METERS +
+                                                        BALL_MAX_RADIUS_METERS + 0.1);
+    double ball_position_x = event.common.world.field().friendlyPenaltyMark().x();
+
+    // Adjust number of tactics based on the number of robots available
+    unsigned int num_tactics = event.common.num_tactics;
+    if (num_tactics != away_from_kick_tactics.size())
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/penalty_kick/penalty_kick_play_fsm.cpp]: doesn't look like we use away_from_kick_tactics (we never add anything to it)

Comment (germchow) [reply src/software/ai/hl/stp/play/penalty_kick/penalty_kick_play_fsm.cpp]: You're right, I actually don't need it

**Review by JoshuaWu7 — APPROVED** [review]

**Review by itsarune — APPROVED** [review]
nice work!

**Pull request merged**